# Python Core Syntax and Environment Setup

## Professional Python for Data Analysis and Machine Learning

---

### Notebook Overview

This notebook provides a comprehensive introduction to the core syntax of Python and the essential tools required to build a professional Python development environment.

The concepts covered here form the foundation for every topic that follows in Data Analysis, Machine Learning, Artificial Intelligence, and general Python development. A strong understanding of these fundamentals is essential before working with libraries such as NumPy, Pandas, Matplotlib, Scikit-learn, TensorFlow, or PyTorch.

The objective of this notebook is not only to learn Python syntax, but also to develop professional coding habits, understand best practices, and build a solid foundation for writing clean, efficient, and maintainable Python programs.

---

## Learning Objectives

After completing this notebook, you will be able to:

- Work confidently with Python variables and built-in data types.
- Understand and use Python operators effectively.
- Perform user input and output using professional formatting techniques.
- Create and manage isolated Python environments using virtual environments.
- Install and manage packages using `pip`.
- Understand the purpose and usage of `requirements.txt`.
- Execute Python programs using different development environments.
- Work efficiently with the Python REPL, Jupyter Notebook, and Google Colab.
- Apply professional coding practices that prepare you for Data Analysis and Machine Learning projects.

---

## Topics Covered

### 1. Variables and Data Types

- Variables
- `int`
- `float`
- `complex`
- `bool`
- `str`
- `None`
- `type()`
- `isinstance()`
- Type Casting
  - `int()`
  - `float()`
  - `str()`

---

### 2. Operators

- Arithmetic Operators
- Comparison Operators
- Logical Operators
- Assignment Operators
- Membership Operators
- Identity Operators

---

### 3. Input and Output

- `print()`
- `input()`
- Formatted String Literals (f-Strings)
- `str.format()`
- Percent (`%`) Formatting

---

### 4. Virtual Environments

- What Virtual Environments Are
- Why They Matter
- Creating Virtual Environments (`venv` and `conda`)
- Activating and Deactivating Environments
- Installing Packages with `pip`
- Creating and Using `requirements.txt`
- Best Practices for Package Management

---

### 5. Running Python

- Python REPL (Interactive Shell)
- Running Python Scripts (`.py`)
- Working with Jupyter Notebook
- Working with Google Colab

---

## Prerequisites

No prior Python knowledge is required. This notebook is designed as the starting point of a complete Python mastery roadmap for Data Analysis and Machine Learning.

---

## Target Audience

This notebook is intended for:

- Python Beginners
- Data Analysts
- Aspiring Machine Learning Engineers
- Data Science Students
- Professionals transitioning into Python-based development

---

## Author

**Abdullah Waseem**

---

## Learning Philosophy

This notebook emphasizes understanding over memorization.

Each topic is explored with practical examples, professional best practices, common mistakes, real-world applications, and connections to Data Analysis and Machine Learning. The goal is to develop a strong conceptual foundation that supports advanced topics throughout the Python learning journey.

# 1. Variables and Data Types

Variables and data types are the foundation of Python programming. Every Python program stores, manipulates, and processes data using variables and different data types.

Python is a dynamically typed language, meaning variables are not restricted to a single data type. Instead, a variable can refer to objects of different types during program execution.

A solid understanding of variables and data types is essential before learning control flow, functions, object-oriented programming, Data Analysis, or Machine Learning.

---

# 1.1 Variables

## What is a Variable?

A variable is a **name (identifier)** that refers to an object stored in memory.

Instead of repeatedly using literal values throughout a program, we assign them to meaningful names. This makes code easier to read, maintain, and modify.

Unlike many programming languages, Python does not require variables to be declared before use. A variable is created automatically when a value is assigned to it.

---

## Syntax

```python
variable_name = value
```

The assignment operator (`=`) assigns the object on the right-hand side to the variable on the left-hand side.

General form:

```python
name = "Abdullah"
age = 22
pi = 3.14159
```

---

## How Variables Work

When Python executes an assignment statement:

1. The object on the right-hand side is created (if it does not already exist).
2. The variable name is created.
3. The variable becomes a reference to that object.

For this reason, variables in Python do **not** store values directly. They store references to objects.

Understanding this concept will become important later when studying mutable and immutable objects, object identity, and memory management.

---

## Why Do We Use Variables?

Variables allow us to:

- Store information for later use.
- Reuse values without rewriting them.
- Write dynamic programs.
- Improve code readability.
- Simplify program maintenance.

Nearly every Python program uses variables, making them one of the most fundamental concepts in the language.

---

## Variable Naming Rules

Variable names (identifiers) must follow these rules:

- Must begin with a letter (`A-Z` or `a-z`) or an underscore (`_`).
- Remaining characters may contain letters, digits, and underscores.
- Cannot begin with a digit.
- Cannot contain spaces.
- Cannot contain special characters such as `@`, `#`, `%`, `$`, or `!`.
- Cannot be a Python keyword.
- Variable names are case-sensitive.


In [ ]:
name = "Abdullah"
age = 22
country = "Pakistan"

print(name)
print(age)
print(country)

Abdullah
22
Pakistan


## Valid Variable Names

The following are valid variable names because they follow Python's identifier rules.


In [ ]:
student_name = "Ali"

_marks = 95

price2026 = 199.99

total_sales = 150000

employee_id = 1001

## Invalid Variable Names

The following examples are invalid because they violate Python's naming rules.


In [ ]:
# Invalid examples (do not execute)
'''
2age = 20          # Starts with a digit

student-name = ""  # Hyphen is not allowed

first name = ""    # Spaces are not allowed

class = "Python"   # 'class' is a reserved keyword'''

'\n2age = 20          # Starts with a digit\n\nstudent-name = ""  # Hyphen is not allowed\n\nfirst name = ""    # Spaces are not allowed\n\nclass = "Python"   # \'class\' is a reserved keyword'

## Variable Reassignment

A variable can refer to a different object at any time.

Reassigning a variable does not modify the original object. Instead, the variable is updated to reference a new object.

This behavior is possible because Python is a dynamically typed language.


In [ ]:
score = 75

print(score)

score = 90

print(score)

75
90


## Multiple Variable Assignment

Python allows assigning multiple values in a single statement.


In [ ]:
name, age, country = "Abdullah", 22, "Pakistan"

print(name)
print(age)
print(country)

Abdullah
22
Pakistan


## Assigning the Same Value to Multiple Variables

The same object can be assigned to multiple variables in a single statement.


In [ ]:
x = y = z = 100

print(x)
print(y)
print(z)

100
100
100


## Professional Naming Conventions (PEP 8)

Python follows the PEP 8 style guide for writing readable code.

Professional variable names should:

- Be descriptive.
- Use lowercase letters.
- Separate multiple words with underscores.
- Clearly describe the stored value.

Preferred examples:

```python
customer_name

average_salary

total_sales

student_age

training_data

test_data
```

Avoid vague names such as:

```python
a

abc

x1

temp

data1
```

unless they are used in a very small and obvious scope.

Readable code is easier to understand, debug, maintain, and collaborate on.

---

# 1.2 Numeric Data Types

Numeric data types are used to store numerical values and perform mathematical operations.

Python provides four built-in numeric data types:

- `int` (Integer)
- `float` (Floating-Point Number)
- `complex` (Complex Number)
- `bool` (Boolean)

Although `bool` is primarily used to represent logical values (`True` and `False`), it is implemented as a subclass of the integer type in Python and therefore belongs to Python's numeric data types.

Choosing the appropriate numeric type is important for writing correct, efficient, and readable programs.

## 1.2.1 Integer (`int`)

An integer is a whole number without a decimal point.

Integers can be:

- Positive
- Negative
- Zero

Unlike many programming languages, Python integers have arbitrary precision. This means they are not limited to 32-bit or 64-bit storage and can grow as large as the available system memory allows.

### Syntax

```python
variable_name = integer_value
```

In [ ]:
age = 22

temperature = -10

students = 45

population = 240000000

print(age)
print(temperature)
print(students)
print(population)

22
-10
45
240000000


Integers are commonly used to represent values that do not require fractional parts.

Examples include:

- Age
- Number of students
- Quantity of products
- Population
- Number of records in a dataset

## 1.2.2 Floating-Point Number (`float`)

A float represents a real number that contains a decimal point.

Floating-point numbers are used whenever precision beyond whole numbers is required.

Examples include measurements, percentages, probabilities, and financial calculations.

In [ ]:
price = 199.99

height = 5.9

accuracy = 98.75

pi = 3.141592653589793

print(price)
print(height)
print(accuracy)
print(pi)

199.99
5.9
98.75
3.141592653589793


Floating-point numbers are stored internally using the IEEE 754 floating-point standard.

Because of this representation, some decimal values cannot be represented exactly in binary.

As a result, floating-point arithmetic may occasionally produce results that appear slightly inaccurate.

This is a normal characteristic of floating-point computation and is common across most programming languages.

In [ ]:
print(0.1 + 0.2)

print(0.1 + 0.2 == 0.3)

0.30000000000000004
False


The output may be surprising because floating-point numbers are approximations rather than exact decimal values.

Later, when working with scientific computing and Data Analysis libraries, we will learn techniques for handling floating-point precision appropriately.

## 1.2.3 Complex Number (`complex`)

A complex number consists of two parts:

- Real part
- Imaginary part

Python represents imaginary numbers using the suffix `j`.

General form:

```python
a + bj
```

where:

- `a` is the real part.
- `b` is the imaginary part.

In [ ]:
z1 = 3 + 4j

z2 = -2 + 7j

print(z1)

print(z2)

(3+4j)
(-2+7j)


The real and imaginary components of a complex number can be accessed using the `.real` and `.imag` attributes.

In [ ]:
number = 5 + 8j

print(number.real)

print(number.imag)

5.0
8.0


Complex numbers are commonly used in fields such as:

- Electrical Engineering
- Signal Processing
- Physics
- Scientific Computing

They are less common in traditional Data Analysis but are fully supported by Python.

## 1.2.4 Boolean (`bool`)

A Boolean represents one of two logical values:

- `True`
- `False`

Boolean values are primarily used in comparisons, conditional statements, loops, filtering operations, and logical expressions.

Although `bool` is its own data type, Python internally treats:

- `True` as `1`
- `False` as `0`

This is why Boolean values can participate in arithmetic operations.

In [ ]:
is_student = True

is_logged_in = False

print(is_student)

print(is_logged_in)

True
False


In [ ]:
print(True + True)

print(True + False)

print(False + False)

2
1
0


Since `True` behaves like `1` and `False` behaves like `0`, the previous arithmetic operations produce integer results.

This behavior is frequently used internally by Python and many data-processing libraries.

In [ ]:
passed = True

marks = 85

print(passed)

print(marks < 50)

print(marks > 50)

True
False
True


Boolean values are fundamental to programming because they control decision-making.

They are extensively used in:

- Conditional statements (`if`, `elif`, `else`)
- Loops (`while`)
- Filtering data
- Comparisons
- Logical operations
- Data validation

In Data Analysis, Boolean values are used heavily for filtering rows, creating masks, handling missing values, and applying conditions to datasets.

---

# 1.3 String (`str`)

A string is a sequence of Unicode characters used to represent textual data.

Strings are one of the most frequently used data types in Python and play a central role in Data Analysis, Machine Learning, Web Development, Automation, and general-purpose programming.

Examples of textual data include:

- Names
- Addresses
- Email addresses
- Product names
- File paths
- Messages
- Customer reviews
- Column names in datasets

In Python, every piece of text is represented as a string.

## Creating Strings

Strings can be created using:

- Single quotes (`' '`)
- Double quotes (`" "`)
- Triple single quotes (`''' '''`)
- Triple double quotes (`""" """`)

Single and double quotes are functionally equivalent. The choice between them depends on readability and coding style.

Triple quotes are primarily used for multi-line strings and documentation strings (docstrings).

In [ ]:
first_name = "Abdullah"

last_name = 'Waseem'

print(first_name)

print(last_name)

Abdullah
Waseem


In [ ]:
message = """
Python is an amazing programming language.

It is widely used in Data Analysis,
Machine Learning,
Artificial Intelligence,
and Automation.
"""

print(message)


Python is an amazing programming language.

It is widely used in Data Analysis,
Machine Learning,
Artificial Intelligence,
and Automation.



## Strings are Immutable

Strings are **immutable**, which means their contents cannot be modified after they are created.

Any operation that appears to modify a string actually creates a new string object.

Understanding immutability is important because many Python data types behave differently depending on whether they are mutable or immutable. We will study this concept in greater depth later.

In [ ]:
language = "Python"

language = "Java"

print(language)

Java


In the previous example, the original string `"Python"` was not modified.

Instead, the variable `language` was updated to reference a new string object containing `"Java"`.

## Escape Characters

Escape characters allow special characters to be included inside strings.

They begin with a backslash (`\`).

In [ ]:
print("He said, \"Python is powerful.\"")

print('It\'s a beautiful day.')

print("First Line\nSecond Line")

print("Name\tAge")

He said, "Python is powerful."
It's a beautiful day.
First Line
Second Line
Name	Age


Some commonly used escape characters are:

| Escape Sequence | Description |
|-----------------|-------------|
| `\'` | Single quote |
| `\"` | Double quote |
| `\\` | Backslash |
| `\n` | New line |
| `\t` | Horizontal tab |

## Raw Strings

A raw string tells Python to treat backslashes as ordinary characters instead of escape characters.

Raw strings are created by placing the letter `r` before the opening quotation mark.

They are especially useful when working with Windows file paths and regular expressions.

In [ ]:
path = r"C:\Users\Abdullah\Documents\Python"
print(path)

C:\Users\Abdullah\Documents\Python


Without a raw string, Python would interpret some backslashes as escape sequences.

Using raw strings avoids this problem and improves readability.

## String Length

The built-in `len()` function returns the number of characters in a string.

Every character, including spaces and punctuation, is counted.

In [ ]:
text = "Data Science"

print(len(text))

12


The `len()` function is one of the most frequently used built-in functions in Python and will be used extensively throughout Data Analysis and Machine Learning workflows.

## Empty String

A string can contain zero characters.

Such a string is called an **empty string**.

In [ ]:
empty = ""

print(empty)

print(len(empty))


0


An empty string is still a valid string object.

It is different from `None`, which represents the absence of a value. We will study `None` in the next section.

## Professional Notes

- Strings are enclosed within quotation marks.
- Strings store textual (Unicode) data.
- Strings are immutable.
- Triple quotes are useful for multi-line text and documentation.
- Raw strings are commonly used for file paths and regular expressions.
- The `len()` function returns the total number of characters in a string.
- An empty string is a valid string object with a length of zero.

---

# 1.4 `None` Type

The `None` type represents the **absence of a value** or the **absence of an object**.

Python provides a single built-in object to represent this concept: `None`.

The data type of `None` is called `NoneType`.

Unlike other data types, there is only **one** `None` object in Python. Every variable assigned `None` refers to this same object.

`None` is commonly used to indicate that:

- A variable has not been assigned a meaningful value yet.
- A value is intentionally missing.
- A function does not return a value.
- An object does not currently exist.

Understanding `None` is important because it appears frequently in professional Python code, Data Analysis libraries, APIs, and Machine Learning workflows.

In [ ]:
value = None

print(value)

None


Although `None` is often described as "nothing" or "no value," it is actually a valid Python object.

It has its own data type (`NoneType`) and can be assigned to variables just like any other object.

In [ ]:
result = None

user = None

data = None

print(result)

print(user)

print(data)

None
None
None


## When Is `None` Used?

`None` is commonly used when a meaningful value is not yet available.

This allows a variable to exist before it receives its final value.

This is preferable to assigning an arbitrary value such as `0` or an empty string, which may have their own meanings.

In [ ]:
customer_email = None

print(customer_email)

None


In the previous example, `customer_email` exists, but no email address has been assigned yet.

Later in the program, the variable can be updated with an actual value.

In [ ]:
customer_email = None

customer_email = "abdullah@example.com"

print(customer_email)

abdullah@example.com


## `None` Is Not the Same as Other "Empty" Values

`None` has a different meaning from values such as:

- `0`
- `0.0`
- `False`
- `""` (empty string)
- `[]` (empty list)
- `{}` (empty dictionary)

These values all represent actual objects with specific meanings.

`None` specifically represents the absence of a value.

In [ ]:
print(None)

print(0)

print(False)

print("")

print([])

None
0
False

[]


Although some of these values may evaluate to `False` in Boolean expressions, they are not equal to `None` and should not be treated as interchangeable.

## Comparing with `None`

The recommended way to check whether a variable is `None` is to use the identity operator `is`.

```python
variable is None
```

Using `is` checks whether both operands refer to the same object.

Since there is only one `None` object in Python, this is the correct and recommended approach.

Avoid using the equality operator (`==`) when checking for `None`.

In [ ]:
value = None

print(value is None)

True


In [ ]:
value = 100

print(value is None)

False


## Professional Notes

- `None` represents the absence of a value.
- `None` is a built-in singleton object in Python.
- The data type of `None` is `NoneType`.
- `None` is different from `0`, `False`, an empty string, or an empty collection.
- The recommended way to compare with `None` is by using `is` or `is not`.
- `None` is widely used in functions, object-oriented programming, APIs, and Data Analysis libraries to represent missing or unavailable values.

---

# 1.5 Determining an Object's Type

Python is a dynamically typed language, which means the data type of an object is determined automatically at runtime.

In many situations, it is useful to determine the type of an object. For example, a program may need to:

- Validate user input.
- Process different data types differently.
- Debug unexpected behavior.
- Verify that function arguments have the expected type.
- Perform data validation in Data Analysis pipelines.

Python provides two built-in functions for this purpose:

- `type()`
- `isinstance()`

Although both functions are used to work with data types, they serve different purposes and should be used appropriately.

## 1.5.1 The `type()` Function

The `type()` function returns the type of an object.

### Syntax

```python
type(object)
```

The returned value is the class (type) of the object.

In [ ]:
age = 22

print(type(age))

<class 'int'>


In [ ]:
print(type(10))

print(type(3.14))

print(type(True))

print(type("Python"))

print(type(None))

<class 'int'>
<class 'float'>
<class 'bool'>
<class 'str'>
<class 'NoneType'>


Notice that the output is displayed as:

```python
<class 'int'>
```

rather than simply:

```python
int
```

This is because, in Python, data types themselves are objects known as **classes**.

At this stage, you only need to understand that `type()` tells us the type of an object. The concept of classes will be studied later in Object-Oriented Programming.

## Practical Example

In [ ]:
customer_name = "Abdullah"

customer_age = 22

salary = 45000.50

is_student = True

print(type(customer_name))

print(type(customer_age))

print(type(salary))

print(type(is_student))

<class 'str'>
<class 'int'>
<class 'float'>
<class 'bool'>


The `type()` function is commonly used while learning Python, debugging programs, and inspecting unfamiliar objects.

However, in professional Python code, another function is often preferred when checking an object's type.

## 1.5.2 The `isinstance()` Function

The `isinstance()` function checks whether an object belongs to a specified type.

Instead of returning the type itself, it returns a Boolean value:

- `True`
- `False`

### Syntax

```python
isinstance(object, type)
```

In [ ]:
age = 22

print(isinstance(age, int))

True


In [ ]:
print(isinstance(3.14, float))

print(isinstance("Python", str))

print(isinstance(True, bool))

print(isinstance(None, type(None)))

True
True
True
True


If the object belongs to the specified type, `isinstance()` returns `True`.

Otherwise, it returns `False`.

In [ ]:
print(isinstance(100, str))

print(isinstance("Python", int))

print(isinstance(3.14, bool))

False
False
False


## Checking Multiple Types

The second argument of `isinstance()` can also be a tuple of types.

This allows an object to be checked against multiple possible data types in a single statement.

In [ ]:
number = 25
print(isinstance(number, (int, float)))

True


In [ ]:
value = 3.14

print(isinstance(value, (int, float)))

True


This feature is particularly useful when a function accepts more than one valid data type.

## `type()` vs `isinstance()`

Although both functions are related to data types, they have different purposes.

| `type()` | `isinstance()` |
|----------|----------------|
| Returns the type of an object. | Checks whether an object belongs to a specified type. |
| Returns a type object. | Returns `True` or `False`. |
| Mainly used for inspection and debugging. | Mainly used for type checking and validation. |
| Does not perform a Boolean check directly. | Directly returns a Boolean result. |

In professional Python code, `isinstance()` is generally preferred for checking an object's type because it is more flexible and works correctly with inheritance, which will be covered later in Object-Oriented Programming.

## Professional Notes

- Use `type()` when you want to know the exact type of an object.
- Use `isinstance()` when you want to verify that an object belongs to a particular type.
- `isinstance()` returns a Boolean value.
- `type()` returns the object's type.
- `isinstance()` can check multiple types by using a tuple.
- In professional Python code, `isinstance()` is generally preferred over `type()` for type checking.

---

# 1.6 Type Casting

Type casting is the process of converting an object from one data type to another.

Python provides several built-in functions for type conversion. The most commonly used are:

- `int()`
- `float()`
- `str()`

Type casting is useful when:

- Processing user input.
- Cleaning and transforming data.
- Performing mathematical operations.
- Preparing data for analysis and Machine Learning.
- Converting values into a desired format.

Not every conversion is valid. A conversion is only successful if the source object can be meaningfully converted into the target type.

## 1.6.1 The `int()` Function

The `int()` function converts a compatible object into an integer.

### Syntax

```python
int(x)
```

If the conversion is successful, an object of type `int` is returned.

In [ ]:
print(int(25.9))

print(int("100"))

print(int(True))

print(int(False))

25
100
1
0


When converting a floating-point number to an integer, the decimal portion is **discarded** rather than rounded.

This behavior is known as **truncation**.

In [ ]:
print(int(8.99))

print(int(8.01))

print(int(-5.8))

8
8
-5


The string passed to `int()` must represent a valid integer.

Otherwise, Python raises a `ValueError`.

In [ ]:
# Valid

print(int("250"))

# Invalid

# print(int("Python"))

# print(int("25.5"))

250


## 1.6.2 The `float()` Function

The `float()` function converts a compatible object into a floating-point number.

### Syntax

```python
float(x)
```

The returned object has the data type `float`.

In [ ]:
print(float(25))

print(float("25"))

print(float("25.75"))

print(float(True))

print(float(False))

25.0
25.0
25.75
1.0
0.0


Unlike `int()`, the `float()` function accepts both integer strings and decimal strings.

If the string cannot be interpreted as a valid floating-point number, a `ValueError` is raised.

In [ ]:
# Valid

print(float("3.14159"))

# Invalid

# print(float("Python"))

3.14159


## 1.6.3 The `str()` Function

The `str()` function converts an object into its string representation.

### Syntax

```python
str(object)
```

Unlike `int()` and `float()`, the `str()` function can convert almost any Python object into a string.

In [ ]:
print(str(100))

print(str(3.14))

print(str(True))

print(str(None))

100
3.14
True
None


In [ ]:
age = 22

message = "Age: " + str(age)

print(message)

Age: 22


Converting objects to strings is commonly required when displaying output, creating reports, writing log messages, or generating text-based files.

## Common Type Conversions

| Expression | Result | Type |
|------------|--------|------|
| `int(8.9)` | `8` | `int` |
| `float(8)` | `8.0` | `float` |
| `str(100)` | `"100"` | `str` |
| `int(True)` | `1` | `int` |
| `int(False)` | `0` | `int` |
| `float(True)` | `1.0` | `float` |
| `float(False)` | `0.0` | `float` |
| `str(True)` | `"True"` | `str` |
| `str(None)` | `"None"` | `str` |

## Invalid Type Conversions

Not every object can be converted into every data type.

Attempting an invalid conversion raises an exception.

Some common examples include:

| Expression | Reason |
|------------|--------|
| `int("Python")` | The string does not represent an integer. |
| `int("25.8")` | The string represents a floating-point number, not an integer. |
| `float("Hello")` | The string does not represent a valid number. |

When working with external data, always ensure that values are valid before attempting type conversion.

## Type Casting in Data Analysis

Type conversion is one of the most common tasks in Data Analysis.

Examples include:

- Converting numeric strings to integers or floating-point numbers.
- Preparing numerical data for calculations.
- Reading data from CSV files where numbers are stored as strings.
- Converting values before visualization.
- Preparing features for Machine Learning models.

Data cleaning and preprocessing frequently involve converting data from one type to another, making type casting an essential skill for every Data Analyst and Machine Learning Engineer.

## Professional Notes

- Type casting converts an object from one data type to another.
- `int()` converts compatible objects to integers.
- `float()` converts compatible objects to floating-point numbers.
- `str()` converts objects to their string representation.
- Invalid conversions raise exceptions.
- Type casting is extensively used in data cleaning, preprocessing, user input handling, and data transformation.

---
# 2. Operators

Operators are special symbols or keywords that perform operations on one or more operands.

An **operand** is the value or object on which an operator acts.

Python provides several categories of operators for performing arithmetic calculations, comparisons, logical evaluations, assignments, membership tests, and identity checks.

A solid understanding of operators is essential for writing expressions, implementing program logic, manipulating data, and performing calculations in Data Analysis and Machine Learning.

---

# 2.1 Arithmetic Operators

Arithmetic operators perform mathematical calculations on numeric operands.

Python supports the following arithmetic operators:

| Operator | Description |
|----------|-------------|
| `+` | Addition |
| `-` | Subtraction |
| `*` | Multiplication |
| `/` | Division |
| `//` | Floor Division |
| `%` | Modulus (Remainder) |
| `**` | Exponentiation |

In [ ]:
a = 10
b = 3

print(a + b)    # Addition

print(a - b)    # Subtraction

print(a * b)    # Multiplication

print(a / b)    # Division

print(a // b)   # Floor Division

print(a % b)    # Modulus

print(a ** b)   # Exponentiation

13
7
30
3.3333333333333335
3
1
1000


## Division (`/`)

The division operator always returns a value of type `float`, even when the mathematical result is a whole number.

In [ ]:
print(10 / 2)

print(type(10 / 2))

5.0
<class 'float'>


## Floor Division (`//`)

The floor division operator returns the quotient after discarding the fractional part and rounding down to the nearest whole number.

This behavior is particularly important when working with negative numbers.

In [ ]:
print(10 // 3)

print(15 // 4)

print(-10 // 3)

3
3
-4


Notice the result of:

```python
-10 // 3
```

The result is `-4`, not `-3`.

Floor division rounds **toward negative infinity**, not toward zero.

## Modulus (`%`)

The modulus operator returns the remainder after division.

It is commonly used for:

- Checking whether a number is even or odd.
- Determining divisibility.
- Implementing cyclic operations.
- Working with indexing and hashing algorithms.

In [ ]:
print(10 % 3)

print(20 % 5)

print(17 % 4)

1
0
1


In [ ]:
number = 18

print(number % 2 == 0)

True


The previous expression evaluates to `True` because `18` is evenly divisible by `2`.

## Exponentiation (`**`)

The exponentiation operator raises the left operand to the power of the right operand.

General form:

```python
base ** exponent
```

In [ ]:
print(2 ** 3)

print(5 ** 2)

print(9 ** 0.5)

8
25
3.0


In the final example,

```python
9 ** 0.5
```

the exponent `0.5` represents the square root, producing the result `3.0`.

In [ ]:
print(625 ** 0.5)

25.0


## Operator Precedence

Arithmetic operators follow the standard order of mathematical operations.

From highest to lowest precedence:

1. `**` (Exponentiation)
2. Unary `+` and `-`
3. `*`, `/`, `//`, `%`
4. `+`, `-`

Parentheses `()` should be used whenever they improve readability or when a specific order of evaluation is required.

In [ ]:
print(2 + 3 * 4)

print((2 + 3) * 4)

print(2 ** 3 * 2)

14
20
16


Using parentheses makes expressions easier to read and avoids ambiguity, even when operator precedence would produce the correct result.

This is considered a good programming practice.

---

# 2.2 Comparison Operators

Comparison operators compare two operands and evaluate the relationship between them.

The result of every comparison is a Boolean value:

- `True`
- `False`

Comparison operators are fundamental to decision-making in Python and are extensively used in conditional statements, loops, filtering, data validation, and Data Analysis.

Python provides the following comparison operators:

| Operator | Description |
|----------|-------------|
| `==` | Equal to |
| `!=` | Not equal to |
| `<` | Less than |
| `>` | Greater than |
| `<=` | Less than or equal to |
| `>=` | Greater than or equal to |

In [ ]:
a = 10
b = 5

print(a == b)

print(a != b)

print(a > b)

print(a < b)

print(a >= b)

print(a <= b)

False
True
True
False
True
False


## Equality Operator (`==`)

The equality operator checks whether two operands have the same value.

If both values are equal, the result is `True`; otherwise, it is `False`.

In [ ]:
print(10 == 10)

print(10 == 5)

print("Python" == "Python")

print("Python" == "python")

True
False
True
False


String comparisons are **case-sensitive**.

For example:

```python
"Python" == "python"
```

evaluates to `False` because uppercase and lowercase letters are treated as different characters.

## Not Equal Operator (`!=`)

The not equal operator checks whether two operands have different values.

It returns `True` when the values are different and `False` when they are equal.

In [ ]:
print(10 != 5)

print(10 != 10)

print("Apple" != "Orange")

True
False
True


## Relational Operators (`<`, `>`, `<=`, `>=`)

Relational operators compare the relative order of two operands.

They are commonly used when comparing numeric values.

In [ ]:
print(20 > 15)

print(20 < 15)

print(20 >= 20)

print(15 <= 20)

True
False
True
True


Comparison operators can also be used with strings.

Strings are compared **lexicographically**, meaning the comparison is based on the Unicode value of each character from left to right.

In [ ]:
print("Apple" < "Banana")

print("Cat" > "Car")

print("abc" < "abd")

True
True
True


Although string comparison is supported, it is primarily useful for sorting and ordering textual data rather than general text comparison.

## Chained Comparisons

Python allows multiple comparisons to be written in a single expression.

General form:

```python
a < b < c
```

This is equivalent to:

```python
a < b and b < c
```

Chained comparisons improve readability and are commonly used in professional Python code.

In [ ]:
age = 22

print(18 <= age <= 30)

print(0 < age < 100)

True
True


## Professional Notes

- Every comparison operator returns a Boolean value (`True` or `False`).
- Comparison operators can be used with numbers, strings, and many other Python objects.
- String comparisons are case-sensitive and are performed lexicographically.
- Chained comparisons provide a clean and readable way to evaluate multiple conditions.
- Comparison operators are the foundation of conditional statements, loops, filtering operations, and logical expressions.

---

# 2.3 Logical Operators

Logical operators combine or modify Boolean expressions.

They are primarily used to evaluate multiple conditions and control the flow of a program.

Python provides three logical operators:

| Operator | Description |
|----------|-------------|
| `and` | Returns `True` if both expressions are `True`. |
| `or` | Returns `True` if at least one expression is `True`. |
| `not` | Reverses the Boolean value of an expression. |

## Logical `and`

The `and` operator returns `True` only when **both** expressions evaluate to `True`.

If either expression is `False`, the result is `False`.

In [ ]:
print(True and True)

print(True and False)

print(False and True)

print(False and False)

True
False
False
False


In [ ]:
age = 25

print(age >= 18 and age <= 30)

print(age > 30 and age < 40)

True
False


The `and` operator is commonly used when **multiple conditions must all be satisfied**.

## Logical `or`

The `or` operator returns `True` if **at least one** expression evaluates to `True`.

It returns `False` only when both expressions are `False`.

In [ ]:
print(True or True)

print(True or False)

print(False or True)

print(False or False)

True
True
True
False


In [ ]:
day = "Saturday"
print(day == 'Saturday' or day == 'Sunday')
print(day == "Monday" or day == "Tuesday")

True
False


The `or` operator is useful when **any one of multiple conditions** is sufficient.

## Logical `not`

The `not` operator reverses the Boolean value of an expression.

| Original Value | Result |
|---------------|--------|
| `True` | `False` |
| `False` | `True` |

In [ ]:
print(not True)

print(not False)

False
True


In [ ]:
is_logged_in = False

print(not is_logged_in)

True


The `not` operator is commonly used to check whether a condition is **not** satisfied.

## Short-Circuit Evaluation

Python evaluates logical expressions using **short-circuit evaluation**.

- For `and`, if the first expression evaluates to `False`, the second expression is **not evaluated**, because the overall result cannot become `True`.

- For `or`, if the first expression evaluates to `True`, the second expression is **not evaluated**, because the overall result is already `True`.

Short-circuit evaluation improves program efficiency and helps prevent unnecessary computations.

In [ ]:
x = 0

print(x != 0 and 10 / x > 1)

False


In the previous example, the second expression is never evaluated because the first expression is already `False`.

This prevents a division-by-zero error and demonstrates one of the practical benefits of short-circuit evaluation.

## Professional Notes

- Logical operators always evaluate Boolean expressions.
- `and` requires all conditions to be `True`.
- `or` requires at least one condition to be `True`.
- `not` reverses a Boolean value.
- Python uses short-circuit evaluation for `and` and `or`, improving both efficiency and safety.

---

# 2.4 Assignment Operators

Assignment operators assign values to variables.

In addition to the standard assignment operator (`=`), Python provides **augmented assignment operators**, which combine an arithmetic operation with assignment.

Common augmented assignment operators include:

| Operator | Equivalent Expression |
|----------|-----------------------|
| `+=` | `x = x + value` |
| `-=` | `x = x - value` |
| `*=` | `x = x * value` |

These operators make code shorter, clearer, and easier to read.

## Addition Assignment (`+=`)

The `+=` operator adds the right operand to the current value of the variable and stores the result back in the same variable.

In [ ]:
score = 80

score += 10

print(score)

90


The previous statement is equivalent to:

```python
score = score + 10
```

## Subtraction Assignment (`-=`)

The `-=` operator subtracts the right operand from the current value of the variable and stores the updated result.

In [ ]:
balance = 500

balance -= 120

print(balance)

380


The previous statement is equivalent to:

```python
balance = balance - 120
```

## Multiplication Assignment (`*=`)

The `*=` operator multiplies the current value of the variable by the right operand and stores the result back in the variable.

In [ ]:
quantity = 6
quantity *= 5

print(quantity)

30


The previous statement is equivalent to:

```python
quantity = quantity * 5
```

## Why Use Augmented Assignment Operators?

Augmented assignment operators:

- Reduce repetitive code.
- Improve readability.
- Clearly indicate that an existing variable is being updated.
- Are widely used in loops, counters, accumulators, and numerical computations.

## Professional Notes

- Augmented assignment operators combine an arithmetic operation with assignment.
- `+=`, `-=`, and `*=` are functionally equivalent to writing the operation explicitly.
- They improve code readability and are considered standard practice in professional Python code.
```

---

# 2.5 Membership Operators

Membership operators are used to determine whether an object exists within another object, known as a **container**.

Common Python containers include:

- Strings
- Lists
- Tuples
- Sets
- Dictionaries

Python provides two membership operators:

| Operator | Description |
|----------|-------------|
| `in` | Returns `True` if the object exists in the container. |
| `not in` | Returns `True` if the object does not exist in the container. |

## The `in` Operator

The `in` operator checks whether an object is present in a container.

In [ ]:
language = "Python"

print("P" in language)

print("Python" in language)

print("Java" in language)

True
True
False


In [ ]:
numbers = [10, 20, 30, 40, 50]

print(20 in numbers)

print(100 in numbers)

True
False


The `in` operator returns a Boolean value:

- `True` if the object is found.
- `False` otherwise.

## The `not in` Operator

The `not in` operator checks whether an object is **absent** from a container.

In [ ]:
language = "Python"

print("Java" not in language)

print("Python" not in language)

True
False


In [ ]:
numbers = [10, 20, 30]

print(40 not in numbers)

print(20 not in numbers)

True
False


## Common Uses

Membership operators are commonly used for:

- Searching within collections.
- Validating user input.
- Filtering data.
- Checking whether a value exists before performing an operation.

They are used extensively in Data Analysis when working with lists, strings, Pandas Series, DataFrames, and other collection types.

## Professional Notes

- Membership operators always return a Boolean value.
- `in` checks for the presence of an object.
- `not in` checks for the absence of an object.
- Membership operators work with many iterable objects, including strings, lists, tuples, sets, and dictionaries.

---

# 2.6 Identity Operators

Identity operators compare the **identity** of two objects.

Unlike comparison operators, which compare **values**, identity operators compare whether two variables refer to the **same object in memory**.

Python provides two identity operators:

| Operator | Description |
|----------|-------------|
| `is` | Returns `True` if both variables refer to the same object. |
| `is not` | Returns `True` if the variables refer to different objects. |

## Value vs Identity

Understanding the difference between **value** and **identity** is essential.

- `==` compares whether two objects have the same value.
- `is` compares whether two variables refer to the same object.

Although these operators may sometimes produce the same result, they answer different questions.

In [ ]:
a = [1, 2, 3]
b = [1, 2, 3]

print(a == b)

print(a is b)

True
False


In the previous example:

- `a == b` returns `True` because both lists contain the same values.
- `a is b` returns `False` because `a` and `b` refer to two different list objects.

In [ ]:
a = [1, 2, 3]

b = a

print(a == b)

print(a is b)

True
True


Here, both variables refer to the same list object.

Therefore, both the value comparison (`==`) and the identity comparison (`is`) return `True`.

## The `is not` Operator

The `is not` operator checks whether two variables refer to different objects.

In [ ]:
x = [10, 20]

y = [10, 20]

print(x is not y)
print(x != y)

True
False


## When Should `is` Be Used?

In everyday programming, `==` is used for comparing values.

The `is` operator should generally be used only when checking object identity.

A common and recommended use is checking for `None`.

```python
value is None
```

This is the standard practice recommended by the Python documentation.

In [ ]:
value = None

print(value is None)

print(value is not None)

True
False


## Professional Notes

- Identity operators compare objects, not values.
- `==` checks value equality.
- `is` checks object identity.
- `is not` checks whether two variables refer to different objects.
- Use `==` for comparing values.
- Use `is` and `is not` primarily when checking against `None` or when object identity is specifically required.

---
# 3. Input and Output

Input and output are fundamental operations in programming.

- **Input** is the process of receiving data from a user, a file, or another source.
- **Output** is the process of displaying or writing data.

Python provides built-in functions for performing input and output operations. The most commonly used are:

- `print()` — Displays output.
- `input()` — Receives input from the user.

In addition, Python provides several techniques for formatting output, including f-strings, `str.format()`, and `%` formatting, which will be covered later in this section.

---

# 3.1 The `print()` Function

The `print()` function displays one or more objects on the standard output (usually the console or terminal).

It is one of the most frequently used built-in functions in Python and is commonly used for displaying results, debugging programs, and presenting information to users.

## Syntax

```python
print(*objects, sep=' ', end='\n', file=None)
```

## Parameters

| Parameter | Description |
|-----------|-------------|
| `*objects` | One or more objects to display. |
| `sep` | Separator inserted between multiple objects. Default is a single space (`' '`). |
| `end` | String appended after the output. Default is a newline (`'\n'`). |
| `file` | Specifies the output destination. By default, output is written to the standard output. |

## Return Value

The `print()` function always returns `None`.

In [ ]:
print("Hello, World!")

Hello, World!


## Printing Multiple Objects

Multiple objects can be passed to `print()` by separating them with commas.

The objects are automatically separated by the value of the `sep` parameter.

In [ ]:
name = "Abdullah"
age = 22
country = "Pakistan"

print(name, age, country)

Abdullah 22 Pakistan


## The `sep` Parameter

The `sep` parameter specifies the separator inserted between multiple objects.

By default:

```python
sep=' '
```

which inserts a single space between objects.

In [ ]:
print("Python", "SQL", "Power BI")

Python SQL Power BI


In [ ]:
print("Python", "SQL", "Power BI", sep=" | ")

Python | SQL | Power BI


In [ ]:
print("Python", "SQL", "NumPy", "Pandas", sep= " | ")

Python | SQL | NumPy | Pandas


In [ ]:
print(2026, 7, 3, sep="-")

2026-7-3


The `sep` parameter is useful when formatting output without manually concatenating strings.

## The `end` Parameter

The `end` parameter specifies what is printed after the output.

By default:

```python
end='\n'
```

which moves the cursor to the next line after printing.

In [ ]:
print("Python")
print("SQL")

Python
SQL


In [ ]:
print("Python", "SQL", "NumPy", end="   ")
print("SQL")

Python SQL NumPy   SQL


In [ ]:
print("Loading", end="...")
print("Done")

Loading...Done


Changing the value of `end` is useful when controlling how output is displayed across multiple `print()` statements.

## The `file` Parameter

The `file` parameter specifies where the output should be written.

By default, `print()` writes to the standard output (the console).

It can also write directly to a file object.

In [ ]:
with open('output.txt', 'w') as file:
  print("Hello, Python!", file = file)

In the previous example, the output is written to a file named `output.txt` instead of being displayed on the screen.

The `file` parameter is useful when generating reports, saving logs, or exporting program output.

## Professional Notes

- `print()` accepts one or more objects.
- Objects are automatically converted to strings before being displayed.
- `sep` controls the separator between objects.
- `end` controls what is printed after the output.
- `file` redirects output to another destination, such as a file.
- `print()` returns `None`.

---

# 3.2 The `input()` Function

The `input()` function pauses program execution and waits for the user to enter data.

When the user presses **Enter**, the entered data is returned as a string.

## Syntax

```python
input(prompt)
```

The `prompt` argument is optional and is displayed before waiting for user input.

## Parameters

| Parameter | Description |
|-----------|-------------|
| `prompt` | An optional message displayed before accepting input. |

## Return Value

The `input()` function always returns a value of type `str`, regardless of what the user enters.

In [ ]:
name = input("Enter your name: ")

print(name)

Enter your name: Abdullah
Abdullah


In [ ]:
age = input("Enter your age: ")

print(age)

print(type(age))

Enter your age: 23
23
<class 'str'>


Even if the user enters a number, the returned value is still a string.

For example, if the user enters:

```text
22
```

the returned object is:

```python
"22"
```

not:

```python
22
```

This is why type casting is often required after using `input()`.

In [ ]:
age = int(input("Enter your age: "))

print(age)

print(type(age))

Enter your age: 25
25
<class 'int'>


In the previous example, the string returned by `input()` is converted into an integer using `int()` before being stored in the variable.

## Professional Notes

- `input()` always returns a string.
- The `prompt` argument is optional but recommended for better user interaction.
- Use type casting when numerical input is required.
- `input()` is widely used in interactive programs, scripts, and command-line applications.

In [ ]:
while True:
    try:
        age = int(input("Enter your age: "))
        break
    except ValueError:
        print("Invalid input. Please enter an integer.")

print(f"Your age is {age}.")

Enter your age: Twenty Five
Invalid input. Please enter an integer.
Enter your age: 25
Your age is 25.


---

# 3.3 Formatted String Literals (f-Strings)

Formatted String Literals, commonly known as **f-strings**, provide a concise and readable way to embed expressions directly within strings.

Introduced in Python 3.6, f-strings are the recommended approach for string formatting because they are more readable, concise, and efficient than older formatting techniques.

They are extensively used in Data Analysis, Machine Learning, reporting, logging, and general Python development.

## Syntax

```python
f"Text {expression}"
```

or

```python
f'Text {expression}'
```

Any valid Python expression enclosed within curly braces (`{}`) is evaluated, and its result is inserted into the string.

In [ ]:
name = "Abdullah"
age = 22

print(f"My name is {name}.")

print(f"I am {age} years old.")

My name is Abdullah.
I am 22 years old.


## Embedding Expressions

An f-string can evaluate any valid Python expression, not just variables.

In [ ]:
a = 10
b = 5

print(f"{a} + {b} = {a + b}")

print(f"{a} × {b} = {a * b}")

print(f"{a} / {b} = {a / b}")

10 + 5 = 15
10 × 5 = 50
10 / 5 = 2.0


Expressions inside the curly braces are evaluated before the string is displayed.

## Calling Functions Inside f-Strings

Function calls can also be placed inside an f-string.

In [ ]:
language = "python"

print(f"Length: {len(language)}")

print(f"Uppercase: {language.upper()}")

Length: 6
Uppercase: PYTHON


Since any valid Python expression can be used inside an f-string, built-in functions and object methods are commonly used for formatting output.

## Formatting Numbers

f-strings support powerful formatting options for controlling the appearance of numerical values.

This is particularly useful when displaying decimal numbers, percentages, or financial data.

General form:

```python
f"{expression:format_specifier}"
```

In [ ]:
pi = 3.1415926535

print(f"{pi:.2f}")

print(f"{pi:.4f}")

3.14
3.1416


In the format specifier:

- `.2f` displays two digits after the decimal point.
- `.4f` displays four digits after the decimal point.

Such formatting is frequently used in reports, dashboards, and Data Analysis projects.

In [ ]:
accuracy = 98.7564

print(f"Model Accuracy: {accuracy:.2f}%")

Model Accuracy: 98.76%


## Professional Notes

- Always prefix the string with `f` or `F`.
- Expressions are enclosed within `{}`.
- Any valid Python expression can be evaluated inside an f-string.
- f-strings are the recommended formatting method in modern Python.
- They provide excellent readability and performance.

---

# 3.4 The `str.format()` Method

Before the introduction of f-strings, the `str.format()` method was the standard approach for formatting strings.

Although modern Python code generally prefers f-strings, `str.format()` is still encountered in older projects and existing codebases.

Understanding this method is useful for reading and maintaining legacy code.

## Syntax

```python
"string".format(values)
```

In [ ]:
name = "Abdullah"
age = 22

print("My name is {}.".format(name))

print("I am {} years old.".format(age))

My name is Abdullah.
I am 22 years old.


Values are inserted into the string in the same order as the placeholders (`{}`).

## Positional Arguments

Each placeholder can reference a specific argument by its position.

In [ ]:
print("{1} is learning {0}.".format("Python", "Abdullah"))

Abdullah is learning Python.


## Named Arguments

Arguments can also be referenced by descriptive names, improving readability.

In [ ]:
print(
    "Name: {name}, Age: {age}".format(
        name="Abdullah",
        age=22
    )
)

Name: Abdullah, Age: 22


## Formatting Numbers

Like f-strings, `str.format()` supports format specifiers.

In [ ]:
pi = 3.1415926535

print("{:.2f}".format(pi))

print("{:.4f}".format(pi))

3.14
3.1416


## f-Strings vs `str.format()`

| Feature | f-Strings | `str.format()` |
|---------|-----------|----------------|
| Readability | Excellent | Good |
| Performance | Faster | Slightly slower |
| Introduced In | Python 3.6 | Python 2.6 |
| Modern Python | Recommended | Mainly for legacy code |

For modern Python development, **f-strings should be your default choice**. Familiarity with `str.format()` remains valuable because it is still present in many existing codebases.

## Professional Notes

- Prefer f-strings in modern Python projects.
- Learn `str.format()` to understand and maintain older code.
- Both techniques support advanced formatting options.
- f-strings are widely used in Data Analysis, Machine Learning, reporting, logging, and automation due to their readability and simplicity.

---
# 4. Virtual Environments

Modern Python projects rarely consist of only the Python language itself.

Most projects rely on **third-party packages** such as:

- NumPy
- Pandas
- Matplotlib
- Scikit-learn
- TensorFlow
- Requests

These packages are installed separately from Python and often have different versions.

Managing these packages correctly is one of the fundamental responsibilities of a Python developer.

To solve package management problems, Python provides **virtual environments**.

Understanding virtual environments is an essential skill for professional software development, Data Analysis, Machine Learning, automation, and web development.

---

# 4.1 What Is a Virtual Environment?

A **virtual environment** is an isolated Python environment that contains its own:

- Python interpreter
- Installed packages
- Package versions

Each virtual environment operates independently from other environments on the same computer.

This means that packages installed inside one environment do not affect any other project.

Instead of sharing one global collection of packages, every project can have its own isolated environment.

For example:

Project A may require:

- Pandas 2.2
- NumPy 2.0

while Project B may require:

- Pandas 1.5
- NumPy 1.24

Both projects can coexist on the same computer because each uses its own virtual environment.

## Why Is This Necessary?

Imagine you are working on two different projects.

Project A was created two years ago.

Project B is a brand-new Machine Learning project.

Both projects require different versions of the same library.

Without virtual environments, installing the newer version would replace the older version for every project.

As a result, Project A might stop working.

Virtual environments solve this problem by isolating each project's dependencies.

## Without a Virtual Environment

```
Computer
│
├── Python
│
└── Installed Packages
    ├── NumPy
    ├── Pandas
    ├── Matplotlib
    └── Requests
```

Every project shares the same packages.

Updating or removing a package affects every project that uses Python.

## With Virtual Environments

```
Computer
│
├── Project A
│   ├── Source Code
│   └── Virtual Environment
│       ├── NumPy 1.24
│       └── Pandas 1.5
│
├── Project B
│   ├── Source Code
│   └── Virtual Environment
│       ├── NumPy 2.0
│       ├── Pandas 2.2
│       └── Scikit-learn
│
└── Project C
    ├── Source Code
    └── Virtual Environment
        ├── Requests
        └── BeautifulSoup
```

Each project has its own isolated environment and package versions.

## Advantages of Virtual Environments

Virtual environments provide several important benefits:

- Package isolation between projects.
- Independent package versions for each project.
- Reduced dependency conflicts.
- Improved reproducibility.
- Easier collaboration with other developers.
- Cleaner project organization.
- Safer experimentation with new libraries.

For these reasons, virtual environments are considered a standard practice in professional Python development.

## Real-World Example

Suppose you develop a Machine Learning project today using the latest version of Scikit-learn.

One year later, you receive a bug report.

Without a virtual environment, your computer may already contain newer package versions.

The project might no longer behave exactly as it did when it was originally developed.

However, if the project uses its own virtual environment, you can recreate the original environment and obtain the same package versions, making debugging and maintenance much easier.

## Why This Matters in Data Analysis

Every Data Analysis project typically depends on multiple third-party libraries.

For example:

- NumPy
- Pandas
- Matplotlib
- Seaborn
- Scikit-learn

Different projects may require different library versions.

Virtual environments ensure that each project remains stable and reproducible regardless of changes made to other projects on your computer.

## Professional Notes

- A virtual environment is an isolated Python environment.
- Each environment maintains its own installed packages and package versions.
- Changes made inside one virtual environment do not affect other projects.
- Professional Python developers create a separate virtual environment for almost every project.
- Virtual environments improve reproducibility, collaboration, and dependency management.

---

# 4.2 Creating a Virtual Environment

Python includes a built-in module named `venv` for creating virtual environments.

The `venv` module creates an isolated Python environment inside your project directory.

Each virtual environment contains its own:

- Python interpreter
- Installed packages
- Package management tools (`pip`)
- Configuration files

Since `venv` is included with the standard Python installation, no additional software is required.

For most Python projects, `venv` is the recommended and most widely used solution for managing virtual environments.

## Project Structure

A virtual environment should be created **inside the project's root directory**.

A typical project structure looks like this:

```text
My_Project/
│
├── .venv/
├── main.py
├── requirements.txt
└── README.md
```

The `.venv` directory contains the virtual environment.

Everything outside `.venv` belongs to your project.

## Step 1 — Open the Project Folder

1. Open Visual Studio Code.
2. Create or open a project folder.

For example:

```text
Python_Practice/
```

Open this folder in VS Code before creating the virtual environment.

## Step 2 — Open the Integrated Terminal

In Visual Studio Code:

- Select **Terminal → New Terminal**

or use the keyboard shortcut:

- **Ctrl + `** (backtick)

The terminal should open in your project's root directory.

You can verify your current directory by running:

```bash
pwd
```

On Windows PowerShell:

```powershell
Get-Location
```

## Step 3 — Create the Virtual Environment

Run the following command:

```bash
python -m venv .venv
```

### Command Breakdown

- `python` runs the Python interpreter.
- `-m` tells Python to execute a module.
- `venv` is the built-in module used to create virtual environments.
- `.venv` is the name of the virtual environment directory.

You may choose any name for the environment, but `.venv` is the most commonly used naming convention in professional Python projects.

`python -m venv .venv`

If the command completes successfully, a new directory named `.venv` will appear inside your project.

Your project structure will now resemble:

```text
Python_Practice/
│
├── .venv/
│   ├── Include/
│   ├── Lib/
│   ├── Scripts/      (Windows)
│   └── pyvenv.cfg
│
└── main.py
```

The contents of the `.venv` directory are managed automatically by Python and should not be modified manually.

## What Happens Internally?

When the virtual environment is created, Python performs several tasks automatically:

- Creates an isolated copy of the Python environment.
- Installs a separate copy of `pip`.
- Creates directories for future package installations.
- Generates configuration files for the environment.

No third-party packages are installed at this stage.

The environment is initially almost identical to a fresh Python installation.

## Why Use the Name `.venv`?

Although the environment can have any name, such as:

```text
env
myenv
project_env
python_env
```

the name `.venv` is widely recommended because:

- It clearly identifies the directory as a virtual environment.
- Many IDEs, including Visual Studio Code, automatically detect it.
- The leading dot (`.`) indicates that it is an implementation detail of the project rather than source code.

## Common Mistakes

Avoid these common mistakes when creating virtual environments:

- Creating one virtual environment for multiple unrelated projects.
- Creating the virtual environment outside the project directory.
- Deleting files inside the `.venv` directory manually.
- Copying a virtual environment from one computer to another.
- Renaming files inside the `.venv` directory.

## Professional Notes

- Use the built-in `venv` module for creating virtual environments.
- Create one virtual environment for each project.
- Place the virtual environment inside the project's root directory.
- The `.venv` naming convention is recommended for consistency and IDE compatibility.
- The contents of the `.venv` directory are managed automatically and should not be edited manually.

---

# 4.3 Activating and Deactivating a Virtual Environment

After creating a virtual environment, it must be **activated** before it can be used.

Activating a virtual environment tells your operating system and terminal to use the Python interpreter and installed packages from that environment instead of the global Python installation.

While the environment is active:

- Running `python` uses the virtual environment's Python interpreter.
- Running `pip` installs packages into the virtual environment.
- Installed packages remain isolated from other projects.

Every time you begin working on a project, the first step should be to activate its virtual environment.

## Why Activation Is Necessary

Consider the following project structure:

```text
Python_Practice/
│
├── .venv/
├── main.py
└── requirements.txt
```

Although the virtual environment exists, your terminal still uses the **global Python installation** by default.

Until you activate the environment, commands such as:

```bash
python
```

and

```bash
pip
```

will continue using the global installation instead of the project's isolated environment.

## Before Activation

```text
Terminal
    │
    ▼
Global Python
    │
    ▼
Global Packages
```

All Python commands use the global environment.

## After Activation

```text
Terminal
    │
    ▼
Project Virtual Environment (.venv)
    │
    ▼
Project Python
    │
    ▼
Project Packages
```

All Python and `pip` commands now use the project's isolated environment.

## Activating the Environment (Windows)

If you are using Windows PowerShell, run:

```powershell
.venv\Scripts\Activate.ps1
```

If you are using Command Prompt (CMD), run:

```cmd
.venv\Scripts\activate.bat
```

After successful activation, your terminal prompt changes.

For example:

Before activation:

```text
PS C:\Python_Practice>
```

After activation:

```text
(.venv) PS C:\Python_Practice>
```

The prefix:

```text
(.venv)
```

indicates that the virtual environment is currently active.

## Activating the Environment (Linux / macOS)

On Linux or macOS, run:

```bash
source .venv/bin/activate
```

Although the command is different, the purpose is exactly the same: activate the project's virtual environment.

## Verifying the Active Python Interpreter

After activation, it is good practice to verify that Python is using the correct interpreter.

Run:

Windows:

```powershell
Get-Command python
```

Linux / macOS:

```bash
Get-Command python
```

The displayed path should point to the Python executable inside the `.venv` directory rather than the global Python installation.

For example, on Windows the output may resemble:

```text
C:\Python_Practice\.venv\Scripts\python.exe
```

This confirms that the virtual environment is active.

## Deactivating the Environment

When you finish working on the project, you can deactivate the virtual environment.

Run:

```text
deactivate
```

This command works on Windows, Linux, and macOS.

After deactivation:

- The `(.venv)` prefix disappears.
- The terminal returns to using the global Python installation.
- Future `python` and `pip` commands use the global environment until another virtual environment is activated.

## Typical Workflow

A professional Python developer typically follows this workflow each time they work on a project:

```text
Open Project Folder
        │
        ▼
Open Terminal
        │
        ▼
Activate Virtual Environment
        │
        ▼
Install Packages (if needed)
        │
        ▼
Write and Run Code
        │
        ▼
Deactivate Environment
```

## Common Mistakes

Avoid these common mistakes:

- Forgetting to activate the virtual environment before starting work.
- Installing packages before activating the environment.
- Assuming the environment activates automatically in every terminal.
- Ignoring the terminal prompt instead of checking for the `(.venv)` prefix.

## Professional Notes

- Creating a virtual environment does not activate it.
- Always activate the virtual environment before running Python or installing packages.
- Verify activation by checking for the `(.venv)` prefix or by confirming the Python interpreter path.
- Deactivate the environment when you finish working on the project.
- Activation ensures that all Python commands and package installations remain isolated within the project.

---

# 4.4 Installing Packages with `pip`

Most Python projects depend on external libraries that are not included with the standard Python installation.

Examples include:

- NumPy
- Pandas
- Matplotlib
- Scikit-learn
- Requests

Python uses a package manager called **`pip`** to install, update, uninstall, and manage these third-party packages.

The name `pip` originally stood for **"Pip Installs Packages."**

When a virtual environment is activated, `pip` installs packages **only into that environment**, keeping each project's dependencies isolated.

## What Is a Package?

A **package** is a collection of Python modules that provides additional functionality.

For example:

- `numpy` provides numerical computing tools.
- `pandas` provides data analysis tools.
- `matplotlib` provides plotting capabilities.
- `requests` enables HTTP requests.

Instead of writing these libraries from scratch, developers install them using `pip`.

## Checking the Installed Version of `pip`

Before installing packages, you can verify that `pip` is available.

Run:

```bash
pip --version
```

Example output:

```text
pip 25.x.x from C:\...\site-packages\pip
```

This confirms that `pip` is installed and ready to use.

## Installing a Package

To install a package, use the following syntax:

```bash
pip install package_name
```

For example, to install NumPy:

```bash
pip install numpy
```

`pip` downloads the package from the Python Package Index (PyPI) and installs it into the active virtual environment.

## Installing a Specific Version

Sometimes a project requires a particular version of a package.

Use the `==` operator to specify the version:

```bash
pip install numpy==2.3.2
```

This installs exactly version `2.3.2` of NumPy.

## Viewing Installed Packages

To see all packages installed in the current environment, run:

```bash
pip list
```

This displays the package names and their installed versions.

## Displaying Information About a Package

To view detailed information about an installed package, use:

```bash
pip show package_name
```

Example:

```bash
pip show numpy
```

This displays information such as:

- Package name
- Version
- Installation location
- Dependencies

## Updating a Package

To upgrade an installed package to the latest available version:

```bash
pip install --upgrade package_name
```

Example:

```bash
pip install --upgrade numpy
```

## Uninstalling a Package

To remove a package from the active virtual environment:

```bash
pip uninstall package_name
```

Example:

```bash
pip uninstall numpy
```

`pip` will ask for confirmation before uninstalling the package.

## Common `pip` Commands

| Command | Purpose |
|---------|---------|
| `pip --version` | Display the installed version of `pip`. |
| `pip install package_name` | Install a package. |
| `pip install package==version` | Install a specific version of a package. |
| `pip list` | List installed packages. |
| `pip show package_name` | Display information about a package. |
| `pip install --upgrade package_name` | Upgrade a package. |
| `pip uninstall package_name` | Remove a package. |

---

# 4.5 The `requirements.txt` File

A `requirements.txt` file is a plain text file that lists all the Python packages required by a project, along with their versions.

Instead of manually remembering or documenting every installed package, developers use a `requirements.txt` file so that the project's dependencies can be recreated automatically on any computer.

This makes projects **portable**, **reproducible**, and **easy to share**.

## Why Is `requirements.txt` Important?

Consider the following scenario:

You develop a Data Analysis project using:

- NumPy
- Pandas
- Matplotlib

A month later, you send the project to another developer.

Without a `requirements.txt` file, they would not know:

- Which packages are required.
- Which versions were used.
- How to recreate your environment.

As a result, the project may fail to run or behave differently.

A `requirements.txt` file solves this problem by documenting the project's dependencies.

## Project Structure

A professional project typically looks like this:

```text
My_Project/
│
├── .venv/
├── main.py
├── requirements.txt
└── README.md
```

The `requirements.txt` file is placed in the project's root directory, alongside the source code.

## Creating a `requirements.txt` File

After installing the required packages, generate the file using:

```bash
python -m pip freeze > requirements.txt
```

The `pip freeze` command lists all installed packages and their exact versions.

The `>` operator redirects the output into a file named `requirements.txt`.

If your virtual environment contains NumPy, the generated file may look similar to:

```text
numpy==2.3.2
```

If additional packages are installed, they will also appear in the file, one package per line.

## Viewing the File

Open `requirements.txt` using Visual Studio Code.

Example:

```text
numpy==2.3.2
pandas==2.3.1
matplotlib==3.10.5
```

Each line specifies:

- Package name
- Exact version

## Installing Packages from `requirements.txt`

Suppose another developer downloads your project.

Instead of installing each package individually, they can install all required packages with a single command:

```bash
python -m pip install -r requirements.txt
```

Here:

- `install` tells `pip` to install packages.
- `-r` means "read the package list from the specified file."

`pip` automatically installs every package listed in the file.

## Typical Workflow

A professional workflow usually follows these steps:

```text
Create Virtual Environment
            │
            ▼
Activate Virtual Environment
            │
            ▼
Install Required Packages
            │
            ▼
Develop the Project
            │
            ▼
Generate requirements.txt
            │
            ▼
Share the Project
            │
            ▼
Another Developer Creates a Virtual Environment
            │
            ▼
Installs Packages Using requirements.txt
```

## Updating `requirements.txt`

Whenever new packages are installed or existing packages are updated, regenerate the file:

```bash
python -m pip freeze > requirements.txt
```

This ensures that the file accurately reflects the current state of the project's environment.

## `pip list` vs `pip freeze`

Although both commands display installed packages, they serve different purposes.

| Command | Purpose |
|---------|---------|
| `python -m pip list` | Displays installed packages in a human-readable table. |
| `python -m pip freeze` | Produces package entries suitable for a `requirements.txt` file. |

For creating `requirements.txt`, always use:

```bash
python -m pip freeze
```

## Common Mistakes

Avoid these common mistakes:

- Creating `requirements.txt` before installing project dependencies.
- Forgetting to update the file after adding new packages.
- Editing package versions manually without a specific reason.
- Assuming another developer has the same packages installed.

## Professional Notes

- Every Python project should include a `requirements.txt` file.
- Generate the file using `python -m pip freeze > requirements.txt`.
- Install dependencies using `python -m pip install -r requirements.txt`.
- Keep the file updated whenever project dependencies change.
- A `requirements.txt` file ensures that the project can be reproduced consistently on any machine.

---

# 4.6 Never Install Packages Globally

One of the most important habits in professional Python development is:

> **Never install project dependencies into the global Python environment.**

Instead, create and activate a virtual environment for every project before installing any packages.

This ensures that each project remains isolated, reproducible, and independent from every other project on your computer.

## What Is the Global Python Environment?

When Python is installed on a computer, it creates a **global environment**.

If you install a package without activating a virtual environment, the package is installed into this global environment.

As a result, every Python project on your computer can access that package.

## Why Is This a Problem?

Suppose you have two projects.

**Project A** requires:

- `numpy` version `1.26`

**Project B** requires:

- `numpy` version `2.3`

If both projects use the global Python environment, installing one version may replace the other.

This can cause one of the projects to stop working or produce unexpected results.

Virtual environments prevent this problem by giving each project its own isolated package installation.

## Global Installation (Not Recommended)

```text
Computer
│
├── Global Python
│
├── NumPy
├── Pandas
├── Matplotlib
└── Requests
│
├── Project A
├── Project B
└── Project C
```

All projects depend on the same package installations.

Changes made for one project can unintentionally affect every other project.

## Virtual Environment (Recommended)

```text
Computer
│
├── Project A
│   └── .venv
│       ├── NumPy 1.26
│       └── Pandas
│
├── Project B
│   └── .venv
│       ├── NumPy 2.3
│       └── Scikit-learn
│
└── Project C
    └── .venv
        ├── Requests
        └── BeautifulSoup
```

Each project has its own isolated dependencies.

Installing or updating packages in one project does not affect the others.

## The Professional Workflow

Every time you start a new Python project, follow this workflow:

1. Create the project folder.
2. Create a virtual environment.
3. Activate the virtual environment.
4. Install the required packages.
5. Develop the project.
6. Generate a `requirements.txt` file.

Following this workflow from the beginning helps avoid dependency conflicts and makes projects easier to maintain and share.

## Are Global Installations Ever Acceptable?

In general, project dependencies should **not** be installed globally.

However, there are a few exceptions.

Tools that are used across many projects, rather than being part of a specific project, may be installed globally.

Examples include:

- Git
- Visual Studio Code
- Python itself

Some developers also install certain command-line tools globally, but project-specific libraries such as `numpy`, `pandas`, and `scikit-learn` should always be installed inside a virtual environment.

## Common Mistakes

Avoid these common mistakes:

- Installing packages before activating the virtual environment.
- Using the same virtual environment for multiple unrelated projects.
- Assuming that packages installed globally are available in every virtual environment.
- Forgetting to include a `requirements.txt` file when sharing a project.

## Professional Notes

- Create a separate virtual environment for every project.
- Activate the virtual environment before installing packages.
- Install project dependencies only inside the active virtual environment.
- Keep project dependencies documented in `requirements.txt`.
- Avoid relying on the global Python environment for project development.

# 5. Running Python

Python programs can be written and executed in several different environments, each designed for specific use cases.

Understanding when and why to use each environment is an essential skill for every Python developer.

In this section, we will cover:

- Python REPL (Interactive Shell)
- Running Python scripts (`.py` files)
- Jupyter Notebooks
- Google Colab

Each environment serves a different purpose, and professional developers choose the one that best fits the task at hand.

---

# 5.1 Python REPL (Interactive Shell)

The **Python REPL** is an interactive environment where Python code is executed immediately after it is entered.

REPL stands for:

- **R**ead
- **E**valuate
- **P**rint
- **L**oop

The REPL continuously performs these four steps:

1. **Read** the code entered by the user.
2. **Evaluate** the code.
3. **Print** the result (if any).
4. **Loop** back and wait for the next input.

This process continues until the user exits the interpreter.

## Why Use the REPL?

The REPL is ideal for:

- Experimenting with Python code.
- Testing small snippets.
- Verifying expressions.
- Learning new functions.
- Debugging simple problems.

Because code executes immediately, the REPL provides instant feedback without creating a Python file.

## Starting the REPL

Open a terminal or command prompt and run:

```bash
python
```

If Python is installed correctly, you should see output similar to:

```text
Python 3.12.x
>>>
```

The `>>>` prompt indicates that the Python REPL is ready to accept Python code.

## Executing Python Code

Once the REPL is running, you can enter Python statements directly.

Each statement is executed immediately after pressing **Enter**.



```
>>> print("Hello, Python!")

Hello, Python!

>>> 10 + 5

15

>>> name = "Abdullah"

>>> print(name)

Abdullah```



## Exiting the REPL

There are several ways to exit the Python REPL.

**Method 1 (Recommended):**

```python
exit()
```

**Method 2 (Windows):**

Press:

```text
Ctrl + Z
```

then press **Enter**.

**Method 3 (Linux/macOS):**

Press:

```text
Ctrl + D
```

## Advantages

- Immediate execution.
- Excellent for experimentation.
- Useful for testing small pieces of code.
- No need to create a Python file.
- Ideal for beginners and quick calculations.

## Limitations

The REPL is **not** suitable for large programs because:

- Code is not automatically saved.
- Organizing large projects is difficult.
- Reusing previous work is inconvenient.
- Version control is impractical.

For larger applications, Python scripts or Jupyter Notebooks are preferred.

## REPL vs Python Scripts

| Python REPL | Python Script |
|-------------|---------------|
| Interactive | File-based |
| Executes one statement at a time | Executes the entire file |
| Best for experimentation | Best for complete programs |
| Temporary | Saved permanently |

## Professional Notes

- The REPL is an interactive Python interpreter.
- It follows the Read–Evaluate–Print–Loop cycle.
- Use the REPL for quick experiments, testing expressions, and learning Python.
- Avoid using the REPL for large or long-term projects.
- Professional developers frequently use the REPL to verify ideas before incorporating them into their programs.

---

# 5.2 Running Python Scripts (`.py` Files)

A **Python script** is a text file containing Python code and saved with the `.py` extension.

Unlike the Python REPL, a script is stored permanently on disk and can be executed any number of times.

Examples:

```text
main.py
calculator.py
data_cleaning.py
app.py
```

Python executes the statements in a script sequentially, starting from the first line and continuing to the last.

## Why Use Python Scripts?

Python scripts are the standard way to build complete programs.

They are used for:

- Automation scripts
- Data processing pipelines
- Machine Learning applications
- Web applications
- APIs
- Desktop applications
- Command-line tools

Because scripts are saved as files, they can be edited, shared, tested, and managed using version control systems such as Git.

## Creating Your First Python Script

Create a new file named:

```text
main.py
```

Add the following code:

```python
print("Hello, Python!")
print("Running my first Python script.")
```

Save the file inside your project folder.

## Project Structure

A simple project may look like this:

```text
My_Project/
│
├── .venv/
├── main.py
├── requirements.txt
└── README.md
```

The source code is stored in Python files such as `main.py`.

## Running a Python Script from the Terminal

Open the terminal inside the project's root directory.

Run:

```bash
python main.py
```

Python reads the file, executes every statement from top to bottom, and displays the output.

## Example

Suppose `main.py` contains:

```python
name = "Abdullah"

print(f"Welcome, {name}!")
```

Run:

```bash
python main.py
```

Output:

```text
Welcome, Abdullah!
```

## How Python Executes a Script

When you run:

```bash
python main.py
```

Python performs the following steps:

1. Opens the file.
2. Reads the source code.
3. Compiles it into bytecode internally.
4. Executes the code line by line.
5. Terminates when the script finishes.

## Running Scripts Inside Visual Studio Code

Visual Studio Code provides multiple ways to run a Python script.

**Method 1: Terminal (Recommended)**

Open the integrated terminal and run:

```bash
python main.py
```

**Method 2: Run Button**

If the Python extension is installed, click the **Run Python File** button in the editor.

Both methods execute the same script.

## Script vs REPL

| Python Script | Python REPL |
|---------------|-------------|
| Stored in a `.py` file | Runs interactively |
| Code is saved permanently | Code is temporary |
| Executes the entire file | Executes one statement at a time |
| Best for complete programs | Best for experimentation |

## Common Mistakes

Avoid these common mistakes:

- Running the script from the wrong directory.
- Forgetting to save the file before running it.
- Using the REPL for large programs.
- Assuming that running a script automatically activates a virtual environment.

---

# 5.3 Jupyter Notebooks

A **Jupyter Notebook** is an interactive web-based environment that allows you to write, execute, and document Python code in a single document.

Unlike a traditional Python script, a Jupyter Notebook combines:

- Executable Python code
- Formatted text
- Mathematical equations
- Tables
- Images
- Charts
- Visualizations

This makes Jupyter Notebooks one of the most popular tools for Data Analysis, Data Science, Machine Learning, and research.

## Why Jupyter Notebooks?

Data Analysis is an exploratory process.

A data analyst rarely writes an entire program from start to finish.

Instead, they typically:

1. Load data.
2. Inspect the data.
3. Clean the data.
4. Transform the data.
5. Visualize the results.
6. Build models.
7. Analyze the output.

A Jupyter Notebook allows each step to be performed independently while preserving all previous work.

## Notebook Structure

A notebook consists of a sequence of **cells**.

Each cell has a specific purpose.

The two most important cell types are:

- Code Cells
- Markdown Cells

## Code Cells

A **Code Cell** contains executable Python code.

When a code cell is run, the notebook sends the code to the Python kernel for execution.

The result is displayed immediately below the cell.

Example:

```python
name = "Abdullah"

print(f"Welcome, {name}!")
```

Output:

```text
Welcome, Abdullah!
```

## Markdown Cells

A **Markdown Cell** contains formatted documentation rather than executable code.

Markdown is commonly used for:

- Titles
- Headings
- Explanations
- Lists
- Tables
- Notes
- Documentation

Throughout this Python series, we are using Markdown extensively to create a professional notebook that explains every concept alongside the corresponding code.

## Execution Order

One important characteristic of Jupyter Notebooks is that **cells can be executed in any order**.

For example:

```text
Cell 1

x = 10
```

```text
Cell 2

print(x)
```

If Cell 2 is executed before Cell 1, Python raises an error because `x` has not yet been created.

For this reason, it is important to execute notebook cells in a logical order.

## The Python Kernel

A **kernel** is the engine that executes the notebook's Python code.

Every time you run a code cell:

Notebook Cell

↓

Python Kernel

↓

Execution

↓

Output

The kernel also stores all variables, functions, and imported libraries currently in memory.

## Restarting the Kernel

Sometimes you may want to clear the notebook's memory.

Restarting the kernel:

- Removes all variables.
- Removes imported modules from memory.
- Clears the current execution state.

After restarting the kernel, every required cell must be executed again.

## Advantages of Jupyter Notebooks

Jupyter Notebooks provide several important advantages:

- Interactive execution.
- Immediate feedback.
- Rich documentation using Markdown.
- Built-in visualization support.
- Easy experimentation.
- Excellent for presenting analyses.
- Widely used in Data Science and Machine Learning.

## Limitations

Although Jupyter Notebooks are extremely powerful, they are not ideal for every situation.

Some limitations include:

- Execution order can become confusing.
- Large software projects are difficult to organize.
- Merge conflicts in version control systems can be challenging.
- Production applications are typically developed as Python packages rather than notebooks.

## Jupyter Notebook vs Python Script

| Jupyter Notebook | Python Script |
|------------------|---------------|
| Interactive | File-based |
| Cell-by-cell execution | Executes the entire file |
| Combines code and documentation | Primarily code |
| Best for analysis and experimentation | Best for applications and automation |
| Saved as `.ipynb` | Saved as `.py` |

## When Should You Use Jupyter Notebooks?

Jupyter Notebooks are the preferred choice for:

- Learning Python
- Data Analysis
- Exploratory Data Analysis (EDA)
- Data Visualization
- Machine Learning experiments
- Research
- Teaching
- Creating technical reports

For software development, automation, APIs, and production applications, Python scripts are generally more appropriate.

## Professional Notes

- A Jupyter Notebook is an interactive document composed of cells.
- The two primary cell types are Code Cells and Markdown Cells.
- Code Cells execute Python code, while Markdown Cells provide documentation.
- The Python kernel executes code and maintains the notebook's state.
- Jupyter Notebooks are the standard environment for Data Analysis and Machine Learning workflows.

---

# 5.4 Google Colab

**Google Colab (Colaboratory)** is a cloud-based interactive notebook environment developed by Google.

It allows users to write, execute, and share Jupyter Notebooks directly through a web browser without installing Python or Jupyter on their local machine.

Google Colab is built on the Jupyter Notebook format, meaning notebooks are saved with the `.ipynb` extension and are compatible with Jupyter Notebook and JupyterLab.

## Why Use Google Colab?

Google Colab eliminates the need to configure a local Python environment.

It provides:

- A pre-configured Python environment.
- Many popular Data Science libraries already installed.
- Automatic integration with Google Drive.
- GPU and TPU support (when available).
- Easy notebook sharing and collaboration.

This makes Google Colab an excellent platform for learning, experimentation, and Machine Learning development.

## Creating a New Notebook

A new notebook can be created by:

1. Opening Google Colab.
2. Selecting **File → New Notebook**.

Each notebook consists of:

- Code Cells
- Markdown Cells

These behave exactly like those in a standard Jupyter Notebook.

## Running Code

Python code is written inside **Code Cells**.

A cell can be executed by:

- Clicking the **Run** button.
- Pressing **Shift + Enter**.

After execution:

- The output appears below the cell.
- Variables remain stored in memory until the runtime is restarted.

## Runtime

Google Colab executes code on a remote machine called the **runtime**.

The runtime provides:

- Python interpreter
- Memory (RAM)
- CPU
- Optional GPU or TPU

Your notebook sends code to the runtime, which executes it and returns the results.

Unlike a local Jupyter Notebook, the computation happens on Google's servers rather than on your own computer.

## Installing Additional Packages

Although many popular libraries are already installed, additional packages can be installed when needed.

Example:

```python
!pip install package_name
```

For example:

```python
!pip install openpyxl
```

The `!` tells the notebook to execute the command in the system shell instead of the Python interpreter.

## Saving Notebooks

Google Colab automatically saves notebooks to Google Drive.

Notebooks can also be downloaded in multiple formats, including:

- `.ipynb`
- `.py`

This makes it easy to continue working locally or share notebooks with others.

## Advantages

Google Colab offers several advantages:

- No local installation required.
- Accessible from any device with a web browser.
- Free access to computational resources.
- Easy sharing and collaboration.
- Automatic notebook saving.
- Well suited for learning, Data Analysis, and Machine Learning.

## Limitations

Google Colab also has some limitations:

- An internet connection is required.
- Sessions may disconnect after periods of inactivity.
- The runtime is temporary, so manually installed packages must be reinstalled in a new session unless otherwise managed.
- Resource availability may vary depending on usage and account type.

## When Should You Use Google Colab?

Google Colab is an excellent choice for:

- Learning Python
- Data Analysis
- Machine Learning
- Deep Learning
- Educational notebooks
- Sharing notebooks with others
- Working from multiple devices

For projects requiring complete control over the development environment, offline work, or extensive software development, a local Python environment is generally more appropriate.

## Professional Notes

- Google Colab is a cloud-based implementation of Jupyter Notebooks.
- It allows Python code to run without local installation.
- Code executes on a remote runtime provided by Google.
- Many popular Data Science libraries are pre-installed.
- Google Colab is widely used for education, experimentation, and Machine Learning.
- Notebooks created in Google Colab are compatible with Jupyter Notebook because both use the `.ipynb` file format.

---

# Conclusion

This notebook introduced the fundamental concepts required to begin professional Python development. It covered the core syntax of the language, essential data types, operators, input and output operations, virtual environments, package management, and the primary environments used to execute Python programs.

The concepts presented in this notebook form the foundation for all subsequent Python topics. A solid understanding of these fundamentals is essential before progressing to control flow, functions, object-oriented programming, data analysis libraries, and machine learning.

By mastering these concepts, a learner is prepared to write, organize, and execute Python programs using professional development practices.

## Topics Covered

- Variables
- Numeric Data Types (`int`, `float`, `complex`, `bool`)
- Strings (`str`)
- The `None` Type
- `type()` and `isinstance()`
- Type Casting (`int()`, `float()`, `str()`)
- Arithmetic Operators
- Comparison Operators
- Logical Operators
- Assignment Operators
- Membership Operators
- Identity Operators
- `print()`
- `input()`
- f-Strings
- `str.format()`
- Virtual Environments
- Creating and Managing Virtual Environments
- Installing Packages with `pip`
- `requirements.txt`
- Running Python
- Python REPL
- Running Python Scripts
- Jupyter Notebooks
- Google Colab